opts/filters.go

```diff
@@ -0,0 +1,98 @@
+package dhtopts
+
+import (
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+// PublicQueryFilter returns true if the peer is suspected of being publicly accessible
+func PublicQueryFilter(h host.Host, ai peer.AddrInfo) bool {
+	if len(ai.Addrs) == 0 {
+		return false
+	}
+
+	var hasPublicAddr bool
+	for _, a := range ai.Addrs {
+		if isRelayAddr(a) {
+			return false
+		}
+		if manet.IsPublicAddr(a) {
+			hasPublicAddr = true
+		}
+	}
+	return hasPublicAddr
+}
+
+// PublicRoutingTableFilter allows a peer to be added to the routing table if the connections to that peer indicate
+// that it is on a public network
+func PublicRoutingTableFilter(conns []network.Conn) bool {
+	for _, c := range conns {
+		addr := c.RemoteMultiaddr()
+		if !isRelayAddr(addr) && manet.IsPublicAddr(addr) {
+			return true
+		}
+	}
+	return false
+}
+
+// PrivateQueryFilter returns true if the peer is suspected of being accessible over a shared private network
+func PrivateQueryFilter(h host.Host, ai peer.AddrInfo) bool {
+	conns := h.Network().ConnsToPeer(ai.ID)
+	if len(conns) > 0 {
+		for _, c := range conns {
+			if manet.IsPrivateAddr(c.RemoteMultiaddr()) {
+				return true
+			}
+		}
+		return false
+	}
+
+	if len(ai.Addrs) == 0 {
+		return false
+	}
+
+	var hasPrivateAddr bool
+	for _, a := range ai.Addrs {
+		if manet.IsPublicAddr(a) {
+			if !isRelayAddr(a) {
+				return false
+			}
+		} else {
+			hasPrivateAddr = true
+		}
+	}
+
+	return hasPrivateAddr
+}
+
+// PrivateRoutingTableFilter allows a peer to be added to the routing table if the connections to that peer indicate
+// that it is on a private network
+func PrivateRoutingTableFilter(conns []network.Conn) bool {
+	for _, c := range conns {
+		if manet.IsPrivateAddr(c.RemoteMultiaddr()) {
+			return true
+		}
+	}
+	return false
+}
+
+// taken from go-libp2p/p2p/host/relay
+func isRelayAddr(a ma.Multiaddr) bool {
+	isRelay := false
+
+	ma.ForEach(a, func(c ma.Component) bool {
+		switch c.Protocol().Code {
+		case ma.P_CIRCUIT:
```

Comment:
could `ma.ValueForProtocol(ma.P_CIRCUIT)` replace this ForEach?

Reply:
should be able to, I wasn't sure if we were trying to avoid using that function or not. While it's here might as well use it though, we can always deprecate and fix it later.

Reply:
That's probably the simplest here. I usually avoid it but _here_, we just care if the address includes a relay component.